pkg/gardenlet/controller/managedseed/add.go

```diff
@@ -0,0 +1,327 @@
+	seed, ok := obj.(*gardencorev1beta1.Seed)
+	if !ok {
+		return false
+	}
+
+	managedSeed := &seedmanagementv1alpha1.ManagedSeed{}
+	if err := p.reader.Get(p.ctx, kutil.Key(p.gardenNamespace, seed.Name), managedSeed); err != nil {
+		return false
+	}
+
+	return filterManagedSeedHelper(p.ctx, p.reader, managedSeed, p.gardenNamespace, p.seedName)
+}
+
+func filterManagedSeedHelper(ctx context.Context, reader client.Reader, managedSeed *seedmanagementv1alpha1.ManagedSeed, gardenNamespace, seedName string) bool {
+	if managedSeed.Spec.Shoot == nil || managedSeed.Spec.Shoot.Name == "" {
+		return false
+	}
+
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := reader.Get(ctx, kutil.Key(managedSeed.Namespace, managedSeed.Spec.Shoot.Name), shoot); err != nil {
+		return false
+	}
+
+	if shoot.Spec.SeedName == nil {
+		return false
+	}
+
+	if shoot.Status.SeedName == nil || *shoot.Spec.SeedName == *shoot.Status.SeedName {
+		return *shoot.Spec.SeedName == seedName
+	}
+
+	return *shoot.Status.SeedName == seedName
+}
+
+// MapSeedToManagedSeed is a mapper.MapFunc for mapping a Seed to the owning ManagedSeed.
+func (r *Reconciler) MapSeedToManagedSeed(_ context.Context, _ logr.Logger, _ client.Reader, obj client.Object) []reconcile.Request {
+	return []reconcile.Request{{NamespacedName: types.NamespacedName{Namespace: r.GardenNamespaceGarden, Name: obj.GetName()}}}
+}
+
+func reconcileRequest(obj client.Object) reconcile.Request {
+	return reconcile.Request{NamespacedName: types.NamespacedName{
+		Name:      obj.GetName(),
+		Namespace: obj.GetNamespace(),
+	}}
+}
+
+var (
+	// RandomDurationWithMetaDuration is an alias for `utils.RandomDurationWithMetaDuration`.Exposed for unit tests.
+	RandomDurationWithMetaDuration = utils.RandomDurationWithMetaDuration
+)
+
+// EnqueueWithJitterDelay returns handler.Funcs which enqueues the object with a random Jitter duration when the JitterUpdate
+// is enabled in ManagedSeed controller configuration.
+// All other events are normally enqueued.
+func (r *Reconciler) EnqueueWithJitterDelay(cfg config.ManagedSeedControllerConfiguration) handler.EventHandler {
+	return &handler.Funcs{
+		CreateFunc: func(evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+			managedSeed, ok := evt.Object.(*seedmanagementv1alpha1.ManagedSeed)
+			if !ok {
+				return
+			}
+
+			generationChanged := managedSeed.Generation != managedSeed.Status.ObservedGeneration
+
+			// Managed seed with deletion timestamp and newly created managed seed will be enqueued immediately.
+			// Generation is 1 for newly created objects.
+			if managedSeed.DeletionTimestamp != nil || managedSeed.Generation == 1 {
+				q.Add(reconcileRequest(evt.Object))
+				return
+			}
+
+			if generationChanged {
+				if *cfg.JitterUpdates {
+					q.AddAfter(reconcileRequest(evt.Object), RandomDurationWithMetaDuration(cfg.SyncJitterPeriod))
+				} else {
+					q.Add(reconcileRequest(evt.Object))
+				}
+			} else {
```

Comment:
```suggestion
				}
				return
			}
```

You could exit early here to reduce the nesting